### ISSUE TITLE
Automatically set sugars or saturated-fat to 0 when carbohydrates or fat is 0

### ISSUE THREAD

**Issue by stephanegigandet** [openfoodfacts/openfoodfacts-server#4561]
### What
- It would help to compute the nutriscore for some products.
e.g. https://world.openfoodfacts.org/product/8888090781045/sunflower-oil-golden-circle

### Part of
- #5538

**Comment by CharlesNepote** [comment]
As of 2022-10-25, [more than 44,000 products are in this case](http://mirabelle.openfoodfacts.org/products?sql=select+*+FROM%0D%0A%28select+count%28*%29+as+all_products+from+%5Ball%5D%29%2C%0D%0A%28select+count%28*%29+as+sat_or_sugar_empty_while_0_fat_or_carb+from+%5Ball%5D+where+%28fat_100g+%3D%3D+0+and+%5Bsaturated-fat_100g%5D+is+null%29+or+%28carbohydrates_100g+%3D%3D+0+and+%5Bsugars_100g%5D+is+null%29%29%2C%0D%0A%28select%0D%0A++%28round%28%0D%0A++100.0*%28select+count%28*%29+from+%5Ball%5D+where%0D%0A+++%28fat_100g+%3D%3D+0+and+%5Bsaturated-fat_100g%5D+is+null%29+or+%28carbohydrates_100g+%3D%3D+0+and+%5Bsugars_100g%5D+is+null%29%0D%0A++%29%0D%0A++%2F%0D%0A++%28select+count%28*%29+from+%5Ball%5D%29%0D%0A++%2C2%29%29+as+percent%0D%0A%29%0D%0A%0D%0A), representing **1.7% of all products**.

**Comment by CharlesNepote** [comment]
A question related to this issue is the interest to know if the 0 is written on the product or computed.

I find interesting to know:
* what is actually printed on the product or not (we can use the hyphen "-" to tell that the value is not on the packaging)
* what is computed or not
But that makes thing more complicated.

Some other nutrient values could be computed based on some rules. For example, if a product is virgin olive oil and he does not have some nutrient values, the missing values could be computed/deduced from either category means or other databases such as CIQUAL, etc.

**Comment by alexgarel** [comment]
@stephanegigandet: 

* I imagine I can do this akin to fix_salt_equivalent (Food.pm) which is called in analyze_and_enrich_product_data (Products.pm) ? 

  I would write a fix_zero_carbohydrates_sugar and fix_zero_fat_saturated

* One question I have is: if carbohydrates is 0 but I have a value for sugar (and same for fat), should I keep the wrong value ? (I would say yes ?).
* More over, should I put "-" if I automatically fix sugar/saturated fats ? It's seems better to me, than putting a 0. (I would verify with a test that nutrition score takes it as 0 in this case)

**Comment by stephanegigandet** [comment]
@alexgarel in fact I think the safest thing would be to not set values to 0, but instead to assume they are 0 when we compute the Nutri-Score.

If carbohydrates is 0 but sugars is not 0, we should not do anything I think, there's no way to know which value is correct or not.

Putting "-" for sugar is not a good idea. "-" indicates that the value is not on the package. But we have no way to know that, it could just be that the value was not entered.

So in short, I think we should just change Nutriscore.pm to assume sugars is 0 when we don't have a sugars value but we have a carbohydrates value.

**Comment by CharlesNepote** [comment]
There are many cases where we could benefit to have a `xxx_computed` or `xxx_normalized` distinguished from the values entered by users:
* polyols and erythritol
* carbs and polyols-sugar-erythritol
* etc.

The logic of these fields would be:
* only compute when the data is missing; when the data have been filed by the user, just copy the data(?)
* only compute when possible: when too many data is lacking, don't compute; it should be different from an estimation
* do not keep the computed value if it throws a data-quality-error
* do not keep the computed value if there is a coherence issue; eg. carb is 0 but sugar is not 0

**Comment by stephanegigandet** [comment]
We could have *_computed and *_estimated indeed. It might be useful to think a bit more about it. The quick and dirty solution would be to suffix the fields and add them to the nutriments hash, but it's already very messy.Maybe we could put a bit more structure.

e.g.

nutrients->{as_sold|prepared}{per 100g|per serving}{listed|computed|estimated}{carbohydrates}

**Comment by stephanegigandet** [comment]
@alexgarel As a short term solution, I think we could create a Nutrients.pm module (to keep Food.pm from growing) with a compute_nutrients($nutrients_ref) function that returns computed nutrients (with things like sugars and saturated fat set to 0).

Then we call it when we create the input to compute the Nutriscore, to check nutrients data quality etc.

Then we can take a bit more time to see if we should also store the result in the product data, and how to do it.

**Comment by alexgarel** [comment]
For the moment I decided to tweak the nutriscore_data structure.

**Comment by alexgarel** [comment]
It's done on #7947 - waiting for reviews.
